### PR TITLE
Tab completion improvement/rewrite

### PR DESCRIPTION
**Pull request by ChaoticMind**
Tab completion can now properly handle elements (most likely usernames) with whitespace. Previously, if someone had a nickname "I like lamp", and one types "I li+tab", it wouldn't work. Another pet peeve tackled!

I'm not super satisfied with it because I would have liked to avoid rebuilding `relevantSubstrings`, but I didn't really find a more elegant solution. (Suggestions welcome). It might be impossible though because of the immutability of strings and the not-worthiness of having a "startsWith()" helper function that works on char arrays.
It's O(n*m) (worst case) where m is the average length of the possible elements (nicknames).

I rewrote a good section of the class to increasing modularity and reduce the need to handle corner cases (although some remnants linger). Commands and Names are now in different `TabCompletionLogic` objects.

### CODE REVIEW

**Comment by ChaoticMind** [comment]
This class is a nice candidate to write tests on because it has some annoying corner cases that we want to make sure are still working if modifications are made (e.g. one character tab completion, input longer than longest tab completion object, etc). And it's a fairly isolated object with simple string inputs/outputs. I couldn't however figure out how to set up the tests with mono (cmdline linux), any tips?
@RoosterDragon, maybe you can help?

**Comment by RoosterDragon** [comment]
I don't think we support that yet sadly: #8849.

**Comment by Mailaender** [comment]
Better don't run it on the command line. You are not a CI bot. Use @MonoDevelop: https://github.com/OpenRA/OpenRA/issues/8849#issuecomment-126804827

File: OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs [hunk 1]
```diff
@@ -19,58 +19,83 @@ public class TabCompletionLogic
...
+		{
+			get
+			{
+				return universe;
+			}
+
+			set
+			{
+				universe = value;
+				longestElementLength = universe.Aggregate("", (max, cur) => max.Length > cur.Length ? max : cur).Length;
```
Comment (RoosterDragon) [inline OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs]: `universe.Max(element => element.Length)` should suffice.

File: OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs [hunk 2]
```diff
@@ -19,58 +19,83 @@ public class TabCompletionLogic
...
 		int currentCandidateIndex = 0;
 		string lastCompleted;
+		public readonly string PrefixOnStart;
+		public readonly string SuffixOnStart;
 		string prefix;
 		string suffix;
 
-		public IList<string> Commands { get; set; }
+		private int longestElementLength;
+		private IList<string> universe;
```
Comment (RoosterDragon) [inline OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs]: Our style guide says the default accessibility modifier should be omitted. i.e. You should remove the explicit `private` here.

File: OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs [hunk 3]
```diff
@@ -19,58 +19,83 @@ public class TabCompletionLogic
...
+
+			set
+			{
+				universe = value;
+				longestElementLength = universe.Aggregate("", (max, cur) => max.Length > cur.Length ? max : cur).Length;
+			}
+		}
 
-		public IList<string> Names { get; set; }
+		public TabCompletionLogic(string prefixOnStart = "", string suffixOnStart = ": ")
```
Comment (RoosterDragon) [inline OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs]: I feel like the empty string would be a better default for the suffix, too.

Comment (Phrohdoh) [inline OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs]: Most clients I've used either append ": " or ", " so one of those two is what I vote for.

**Comment by atlimit8** [comment]
> I couldn't however figure out how to run the current tests with mono (command line on Linux)

See #8861 (reopened)

**Comment by atlimit8** [comment]
> Better don't run it on the command line. You are not a CI bot. Use @MonoDevelop: #8849 (comment)

@Mailaender and what would you put in `git rebase -i --exec 'make ...' bleed` to run the unit tests?

File: OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs (hunk 3 above)
Comment (RoosterDragon) [inline OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs]: To be clear, I just mean the default for `TabCompletionLogic` constructor should be an empty string. The caller should explicitly pass the suffix they're after.

**Comment by ChaoticMind** [comment]
Rebased and updated (bugfix)

Covers the following cases, for both starting a string with a nick or not:

1. nicks with and without space
2. nicks with common substring (space or not)
3. nicks with repeating substring (take longest match) (also both with and without space)

Example for the third point, if two nicks are "cabcabal" and "cabal", typing "cabca + tab" will only match the first nick. This is important to cover because it would otherwise get confusing in cases of nicks with spaces.

I had to keep track of "longest match" for each hit, which I did in a Tuple<string, int> (the `string` is the proposed candidate/nick, and the `int` is the length of the longest common substring between the candidate and the input text). I then filtered on the longest match (largest second element).

Another way would be a dictionary of lists, indexed by the "longest common substring length" and then only take the list for the highest index. Unsure if that would be more readable, probably equivalent.

File: OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs [hunk 4]
```diff
@@ -20,60 +20,91 @@ public class TabCompletionLogic : ChromeLogic
...
 		string lastCompleted;
+		public readonly string PrefixOnStart;
+		public readonly string SuffixOnStart;
 		string prefix;
 		string suffix;
 
-		public IList<string> Commands { get; set; }
+		int longestElementLength;
+		IList<string> universe;
+		public IList<string> Universe
```
Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs]: What is the significance of this name?

Comment (ChaoticMind) [inline OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs]: Set theory [nomenclature](https://en.wikipedia.org/wiki/Universe_%28mathematics%29), seems appropriate.

Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs]: This is potentially confusing here, since we are simulating `Actor`s, and `World`s.  Long ago we were going to have a `Universe` class that wrapped the concept of the world and other bits.

Comment (atlimit8) [inline OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs]: `Pool` or `Completions` should work.

Comment (pchote) [inline OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs]: `Completions` sounds good to me.  It is descriptive and unambigous.

Comment (ChaoticMind) [inline OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs]: Sure thing, I'll wait for the rest of the review and change it.

File: OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs [hunk 5]
```diff
@@ -20,60 +20,91 @@ public class TabCompletionLogic : ChromeLogic
...
+		}
 
 		public string Complete(string text)
 		{
-			if (string.IsNullOrWhiteSpace(text))
+			if (string.IsNullOrEmpty(text) || Universe == null)
 				return text;
 
-			if (lastCompleted == text && candidates.Any())
+			// check if tab is cycled without text modification
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs]: Start comments with an uppercase letter, please.

**Comment by ChaoticMind** [comment]
Updated, rebased, and squashed.

File: OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs [hunk 6]
```diff
@@ -20,60 +20,91 @@ public class TabCompletionLogic : ChromeLogic
...
+			return lastCompleted;
+		}
+
+		void GenerateCompletionCandidates(string text)
+		{
+			suffix = "";
+			candidates = new List<string>();
+			var relevantSubstrings =
+				from i in Enumerable.Range(1, Math.Min(text.Length, longestElementLength))
+				select text.Substring(text.Length - i);
```
Comment (Mailaender) [inline OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs]: > [17:40] pchote: functional syntax, please

**Comment by Mailaender** [comment]
Works as promised. :white_check_mark:

File: OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs (hunk 6 above)
Comment (ChaoticMind) [inline OpenRA.Mods.Common/Widgets/Logic/TabCompletionLogic.cs]: fixed

**Comment by RoosterDragon** [comment]
Any chance of some unit tests for this now  #8861 is merged?

**Comment by atlimit8** [comment]
Considering the concern above, it would be nice for the unit tests to be added now rather than later or possibly never. @ChaoticMind use `make nunit` to run NUnit tests in a Linux terminal.

**Comment by abcdefg30** [comment]
Needs a rebase.

**Comment by obrakmann** [comment]
Looks good for the most part. I found one issue: when you type "/og", for example, and press tab, it will auto-complete to "/ogiveall", which is obviously wrong.

**Comment by Phrohdoh** [comment]
@ChaoticMind Are you still working on this?

**Comment by ChaoticMind** [comment]
I haven't worked on this since last commit. It's still on my radar but I can't do it before early February. If you prefer to close it and reopen it when I push an update, that's OK for me. Sorry :(

@obrakmann Yeah, I was planning on making that adjustment in a PR after this one (to keep them smaller). But I guess I may as well just include it in this one.

**Comment by pchote** [comment]
Any updates here?

**Comment by reaperrr** [comment]
Since there hasn't been any activity for two months, we're closing this to free the PR queue.
@ChaoticMind fell free to reopen it when you get around to update it.
